rust/blockstore/src/arrow/block/delta/data_record.rs

```diff
@@ -340,110 +239,27 @@ impl DataRecordStorage {
-                    // ok to initialize fixed size float list with fixed size as 0.
-                    embedding_dim = 0;
-                    embedding_builder = FixedSizeListBuilder::new(Float32Builder::new(), 0);
-                    id_builder = StringBuilder::new();
-                    metadata_builder = BinaryBuilder::new();
-                    document_builder = StringBuilder::new();
-                } else {
-                    embedding_dim = storage.iter().next().unwrap().1 .1.len();
-                    // Assumes all embeddings are of the same length, which is guaranteed by calling code
-                    // TODO: validate this assumption by throwing an error if it's not true
-                    let total_embedding_count = embedding_dim * item_capacity;
-                    id_builder = StringBuilder::with_capacity(item_capacity, inner.id_size);
-                    embedding_builder = FixedSizeListBuilder::with_capacity(
-                        Float32Builder::with_capacity(total_embedding_count),
-                        embedding_dim as i32,
-                        item_capacity,
-                    );
-                    metadata_builder =
-                        BinaryBuilder::with_capacity(item_capacity, inner.metadata_size);
-                    document_builder =
-                        StringBuilder::with_capacity(item_capacity, inner.document_size);
-                }
-                for (key, (id, embedding, metadata, document)) in storage.into_iter() {
-                    key_builder.add_key(key);
-                    id_builder.append_value(id);
-                    let embedding_arr = embedding_builder.values();
-                    for entry in embedding {
-                        embedding_arr.append_value(entry);
-                    }
-                    embedding_builder.append(true);
-                    metadata_builder.append_option(metadata.as_deref());
-                    document_builder.append_option(document.as_deref());
-                }
-            }
-            Err(_) => {
-                panic!("Invariant violation: SingleColumnStorage inner should have only one reference.");
-            }
+        let inner = Arc::try_unwrap(self.inner)
+            .expect(
+                "Invariant violation: SingleColumnStorage inner should have only one reference.",
+            )
+            .into_inner();
+        let storage = inner.storage;
+
+        let mut value_builder =
+            <&DataRecord as ArrowWriteableValue>::get_arrow_builder(inner.size_tracker);
+
+        for (key, value) in storage.into_iter() {
+            key_builder.add_key(key);
+            <&DataRecord as ArrowWriteableValue>::append(value, &mut value_builder);
         }
+
         // Build arrow key with fields.
         let (prefix_field, prefix_arr, key_field, key_arr) = key_builder.as_arrow();
+        let (struct_field, value_arr) = <&DataRecord as ArrowWriteableValue>::finish(value_builder);
```

Comment:
This would have been a good place to have b/w compat tests in case we accidentally change the storage.